pkg/server/handlers/access_grant.go

```diff
@@ -0,0 +1,110 @@
+package handlers
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/storacha/go-libstoracha/capabilities/access"
+	"github.com/storacha/go-ucanto/core/dag/blockstore"
+	"github.com/storacha/go-ucanto/core/delegation"
+	"github.com/storacha/go-ucanto/core/invocation"
+	"github.com/storacha/go-ucanto/core/receipt/fx"
+	"github.com/storacha/go-ucanto/core/result"
+	"github.com/storacha/go-ucanto/core/result/failure"
+	"github.com/storacha/go-ucanto/principal"
+	"github.com/storacha/go-ucanto/server"
+	"github.com/storacha/go-ucanto/ucan"
+)
+
+// validity is the time a granted delegation is valid for.
+const validity = time.Hour
+
+func NewAccessGrantHandler(id principal.Signer) server.HandlerFunc[access.GrantCaveats, access.GrantOk, failure.IPLDBuilderFailure] {
+	return func(
+		ctx context.Context,
+		cap ucan.Capability[access.GrantCaveats],
+		inv invocation.Invocation,
+		ictx server.InvocationContext,
+	) (result.Result[access.GrantOk, failure.IPLDBuilderFailure], fx.Effects, error) {
+		nb := cap.Nb()
+		log.Infow(
+			"handling access request",
+			"ability", access.Grant,
+			"issuer", inv.Issuer().DID(),
+			"capabilities", nb.Att,
+			"cause", nb.Cause,
+		)
+		var cause invocation.Invocation
+		if cap.Nb().Cause != nil {
+			bs, err := blockstore.NewBlockReader(blockstore.WithBlocksIterator(inv.Blocks()))
+			if err != nil {
+				return nil, nil, fmt.Errorf("reading invocation blocks: %w", err)
+			}
+			i, err := invocation.NewInvocationView(cap.Nb().Cause, bs)
```

Comment:
nit: can we rename this to `cInv` or something else that's not `i`? Sorry, too many years of `i` being the loop variable.